### PR TITLE
improve CI setup

### PR DESCRIPTION
**Pull request by chenrui333**
- use py3.12 for ci
- bump action dependencies

---

- https://github.com/Homebrew/homebrew-core/pull/165881

cc @wismill

### CODE REVIEW

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -23,11 +23,7 @@ jobs:
           python-version: '3.12'
       - name: Install dependencies
         run: |
-          brew install bison graphviz libxml2 meson ninja pyyaml
-          # HACK: force Doxygen version 1.9.6, because 1.9.7 breaks our cool URIs.
-          curl -O https://raw.githubusercontent.com/Homebrew/homebrew-core/d2267b9f2ad247bc9c8273eb755b39566a474a70/Formula/doxygen.rb
-          brew install --formula doxygen.rb
-          rm doxygen.rb
```
Comment (whot) [inline .github/workflows/macos.yml:31]: fwiw, I think an option here would also be to simply disable the doc build on macos if this ever happens again. I don't think we really need to CI that, especially if it's a bit buggy.

Come to think of it: maybe we should just disable the documentation build by default? I doubt there are (m)any users out there who need a local doxygen result -> #471

**Review by whot — APPROVED** [review]
This LGTM, see my `pip` vs `brew` question though please, thanks.

File: .github/workflows/macos.yml (hunk 1 above)
Comment (wismill) [reply .github/workflows/macos.yml:31]: #471 merged. We can remove Doxygen requirement here

File: .github/workflows/macos.yml [hunk 2]
```diff
@@ -23,8 +23,7 @@ jobs:
           python-version: '3.12'
       - name: Install dependencies
         run: |
-          python -m pip install --upgrade meson PyYAML
-          brew install libxml2 bison ninja graphviz
+          brew install bison graphviz libxml2 meson ninja pyyaml
```
[the remark this answers is not in the document]
Comment (wismill) [reply .github/workflows/macos.yml]: #472 merged. We can remove graphviz requirement.

[the remark this answers is not in the document]
Comment (chenrui333) [reply .github/workflows/macos.yml]: cool, I will remove that

File: .github/workflows/macos.yml (hunk 1 above)
Comment (chenrui333) [reply .github/workflows/macos.yml:31]: 👍

File: .github/workflows/macos.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenrui333) [reply .github/workflows/macos.yml]: brew should be always latest meson by default, although there is some current pending update going on, https://github.com/Homebrew/homebrew-core/pull/165908

**Review by wismill — APPROVED** [review]
Rebased and squashed some commits

**Comment by wismill** [comment]
Thanks!

**Pull request merged**